Colorize zpool status output

tonyhutter opened the pull request:
### Motivation and Context
Add colored `zpool status` output so admins can quickly spot a unhealthy vdev/pool.

### Description
If the `ZFS_COLOR` env variable is set, then use ANSI color output in `zpool status`:

- Column headers are bold
- Degraded or offline pools/vdevs are yellow
- Non-zero error counters and faulted vdevs/pools are red
- The 'status:' and 'action:' sections are yellow if they're displaying a warning. 

**Screenshots:**

![zpool_color3](https://user-images.githubusercontent.com/11469457/65365057-104f7e80-dbcb-11e9-8fda-9699b87f6e6d.png)

![zpool_color2](https://user-images.githubusercontent.com/11469457/65365059-13e30580-dbcb-11e9-8e4b-80f070cb0842.png)

### How Has This Been Tested?
Manually tested.  Added test case to look for bold, yellow and red text.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

tonyhutter commented:
Looks like I'm seeing some test failures:
```
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zpool/zpool_colors (run as root) [00:04] [FAIL]
20:54:29.63 ASSERTION: Test colorized zpool status output
20:54:29.70 SUCCESS: dd if=/dev/urandom of=//mnt/testdir/testfile bs=10M count=1
20:54:30.56 SUCCESS: zpool export testpool
20:54:31.78 SUCCESS: zpool import testpool
20:54:31.78 'unknown': I need something more specific.
20:54:31.78 NOTE:   
20:54:31.82 SUCCESS: zpool offline -f testpool loop4
```
This is coming from ncurses:
```
'unknown': I need something more specific.
```
Maybe we need `ncurses-term` installed on Ubuntu?:
 
https://unix.stackexchange.com/questions/213726/ssh-from-screen-leads-to-unknown-terminal-error

I'm looking into it.

ahrens commented:
Live demo from the 2018 OZDS hackathon, where this project won first place 🥇 : https://youtu.be/zN_tGxCpTBU?t=430

When do we get 🔥 emoji support? :-)

ahrens commented:
Would it be possible to auto-detect if the terminal supports color and turn this on by default, like `git` seems to do?

kpande commented:
no please, use a value in /etc/default/zfs instead and leave it off by default.

tonyhutter commented:
@ahrens  one issue is that people screen scrape the zpool status output with awk.  Each color escape code counts as a field in awk, so if we turned on color by default it's probably going to mess up the column counting in a bunch of scripts.

I decided to pass on the emojis for now, since I didn't know of a good way to detect if they're supported in the terminal.  And they're a little too silly :-)

lundman commented:
Generally, things like git use `isatty()` to have colour, and disable if not, so if you pipe to awk it will be disabled.

spacelama commented:
And --color[=yes|auto|no (default yes if --color supplied without a parameter)], defaulting to
--color=auto if --color not supplied.

tonyhutter commented:
@lundman's thanks for the suggestion about `isatty()`.  Let me look into that.  I'd personally like the colors on by default if we know that it's not going to interfere with anything.

kpande commented:
some of my scripts fake being used in a TTY for one reason or another so it would require me to modify them to disable colour.

richardelling commented:
tty colors are a human factors disaster. theee is no way for the dev to know the contrast of the terminal’s colors. so the dev says “I think good things should be green” without realizing that some folks use green as the background, rendering it unreadable. the typical egregious colors are blue/black, but the real problem is you have no idea what the users preferences are. also cut-n-paste often picks up the background and foreground colors... spreading the ugliness. personally, I use different color backgrounds to differentiate being logged into different machines, workflows, or tasks. but the most telling story is recently a colleague did this in a product, when we went onsite with a customer, the customer’s preferences rendered more than 50% of the text unreadable. thus the product had to be changed to allow the customer to choose color combinations (of which there are thousands to choose from) for each and every possible rendering.  human factors disaster compounded. 

just say no to colors by default

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2033,6 +2033,26 @@ print_status_trim(vdev_stat_t *vs, boolean_t verbose)
+	if (strcmp(health, "FAULTED") == 0 ||
+	    strcmp(health, "SUSPENDED") == 0 ||
+	    strcmp(health, "UNAVAIL") == 0)
+		return (ANSI_RED);
```
nit: usually we add `{ }` for any multi-line conditional.

behlendorf commented on lib/libzfs/libzfs_util.c:
```diff
@@ -46,6 +46,18 @@
+#include <term.h>
+/*
+ * For some reason, term.h #defines 'lines' (in lower case) as a macro.
+ * Since we also have a variable named 'lines', undefine the macro here.
```
Even though this works, I think we should go ahead and rename the our local `lines` variable to avoid the conflict entirely.  Then drop this block.  According to the curses documentation, `curs_terminfo(3X)`.

```
      ·   If use_env(FALSE) has been called, values for lines and  columns  specified
           in terminfo are used.
```

Which would be handy if we decide it's desirable to use more of the libraries functionality.

behlendorf commented on lib/libzfs/libzfs_util.c:
```diff
@@ -1886,3 +1898,101 @@ zfs_version_print(void)
+		 * If it's not supported, then silently don't use it.
+		 */
+		if (setupterm(NULL, fileno(stdout), (int *)0) == 0 &&
+		    tigetnum("colors") >= 16) {
```
Maybe `&& isatty(stdout)` for good measure.

behlendorf commented on lib/libzfs/libzfs_util.c:
```diff
@@ -1886,3 +1898,101 @@ zfs_version_print(void)
+		color_start(color);
+
+	va_start(aptr, format);
+	vsprintf(buf, format, aptr);
```
Can you use `vsnprintf()` here to avoid any overflow and make sure it's null-terminated.

behlendorf commented on man/man8/zpool.8:
```diff
@@ -2867,6 +2867,12 @@ or
+Use ANSI color in
+.Nm zpool status
+output.
+.El
```
Can you move this up after `ZFS_ABORT` to keep the environment variables alphabetically ordered.  Moving` `ZFS_VDEV_DEVID_OPT_OUT` up wouldn't be a bad thing either.

behlendorf commented on tests/zfs-tests/tests/functional/cli_root/zpool/zpool_colors.ksh:
```diff
@@ -0,0 +1,85 @@
+# 3. Look for 'FAULTED' in red
+#
+
+verify_runnable "both"
```
It'd be nice if we could somehow selectively skip this test when not built `--with-ncurses`.  Though I don't have a great suggestion for how to do that.

tonyhutter commented:
It sounds like there's no consensus on colors on/off by default, so we'll keep them off unless you `export ZFS_COLOR=1`.
 
@behlendorf I implemented all your changes in my latest push.  I'm still getting test failures in buildbot though, so I've added some instrumentation printfs.  Please don't pull this in until I can get all the tests passing.

tonyhutter commented:
@behlendorf it's passing the tests now.  I've removed the "WIP" tag.

behlendorf commented on tests/zfs-tests/tests/functional/cli_root/zpool/zpool_colors.ksh:
```diff
@@ -0,0 +1,93 @@
+log_must dd if=/dev/urandom of=/$TESTDIR/testfile bs=10M count=1
+
+log_must zpool export $TESTPOOL
+log_must zpool import $TESTPOOL
```
Why export and import?  `log_must zpool sync` can be used instead to make sure the datas on disk.

behlendorf commented on tests/zfs-tests/tests/functional/cli_root/zpool/zpool_colors.ksh:
```diff
@@ -0,0 +1,93 @@
+log_must zinject -d $DISK2 -e io -T write -f 20 $TESTPOOL
+
+
+log_must zpool scrub $TESTPOOL
```
Now that `zpool wait` has been integrated you can instead use `log_must zpool scrub -w $TESTPOOL` and drop the next `wait_scrubbed` line.

behlendorf commented on tests/zfs-tests/tests/functional/cli_root/zpool/zpool_colors.ksh:
```diff
@@ -0,0 +1,93 @@
+
+# Use 'script' to fake zpool status into thinking it's running in a tty.
+# Log the output here in case it's needed for postmortem.
+log_note "$(script --return --quiet -c 'TERM=xterm-256color ZFS_COLOR=1 zpool status' /dev/null)"
```
nit: > 80 chars

behlendorf commented on tests/zfs-tests/tests/functional/cli_root/zpool/zpool_colors.ksh:
```diff
@@ -0,0 +1,93 @@
+log_note "$(script --return --quiet -c 'TERM=xterm-256color ZFS_COLOR=1 zpool status' /dev/null)"
+
+# Replace the escape codes with "ESC" so they're easier to grep
+out="$(script --return --quiet -c 'TERM=xterm-256color ZFS_COLOR=1 zpool status' /dev/null | \
```
nit: > 80 chars

behlendorf commented on tests/zfs-tests/tests/functional/cli_root/zpool/zpool_colors.ksh:
```diff
@@ -0,0 +1,93 @@
+# Luckily, awk counts the start and end escape codes as separate fields, so
+# we can easily remove the vdev field to get what we want.
+#
+out="$(script --return --quiet -c 'TERM=xterm-256color ZFS_COLOR=1 zpool status' /dev/null \
```
nit: > 80 chars

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2033,6 +2033,26 @@ print_status_trim(vdev_stat_t *vs, boolean_t verbose)
+
+	if (strcmp(health, "OFFLINE") == 0 ||
+	    strcmp(health, "DEGRADED") == 0 ||
+	    strcmp(health, "REMOVED") == 0)
```
nit: can you add { } for this multi-line conditional and the one above.

tonyhutter commented:
@behlendorf my latest push fixes all your comments

behlendorf commented:
@lundman @freqlabs would you mind double checking this PR to ensure it does make sense for your respective platforms.  We can f course tweak it latter if needed.  By default, the colors are disabled so you will need to set the `ZFS_COLOR` environment variable if you test it locally.

lundman commented:
>  PR to ensure it does make sense for your respective platforms.

Looks to be standard porting, `strcmp` -> `strlcmp` for the noise reduction (deprecated) , not sure about `exit()` in a library, but if you are OK with it, so am I.  
Not sure what comment `"FG" for forground color` refers to, but then... it's 'just a comment' :) 

Yeah, should be fine to port over.

freqlabs commented on tests/zfs-tests/tests/functional/cli_root/zpool/zpool_colors.ksh:
```diff
@@ -0,0 +1,94 @@
+
+# Use 'script' to fake zpool status into thinking it's running in a tty.
+# Log the output here in case it's needed for postmortem.
+log_note "$(script --return --quiet -c \
```
The `script` command will be different on FreeBSD, but it seems to still meet the basic need here. :+1: 
```
$ echo $(script -q /dev/null python2.7 -c "import sys; print sys.stdout.isatty())")
True
```
The discrepancy could be hidden behind a function to make it easier to add alternative implementations.

freqlabs replied:
FreeBSD support has landed in ZTS. We don't have FreeBSD bots blocking PRs yet, but let's try to get this addressed now.
I suggest putting the following under the cleanup function:
```
function faketty
{
	if is_freebsd; then
		script -q /dev/null "$@"
	else
		script --return --quiet -c "$*" /dev/null
	fi
}
```
And then use it like so:
```
log_note "$(faketty TERM=xterm-256color ZFS_COLOR=1 zpool status)"
```
```
out="$(faketty TERM=xterm-256color ZFS_COLOR=1 zpool status | \
    grep -E 'pool:|DEGRADED' | \
    sed -r 's/\s+//g;'$(echo -e 's/\033/ESC/g'))"
```
And so on.

tonyhutter replied:
Thanks for the code.  I'll add it in

tonyhutter commented:
@lundman @freqlabs thanks, I'll implement your changes.

freqlabs commented on tests/zfs-tests/tests/functional/cli_root/zpool/zpool_colors.ksh:
```diff
@@ -0,0 +1,91 @@
+
+# Use 'script' to fake zpool status into thinking it's running in a tty.
+# Log the output here in case it's needed for postmortem.
+log_note "$(faketty 'TERM=xterm-256color ZFS_COLOR=1 zpool status')"
```
`TERM=xterm-256color ZFS_COLOR=1 zpool status` needs to be unquoted in these.

tonyhutter replied:
Thanks, I removed the quotes in my latest push

tonyhutter commented:
@behlendorf as a side note - we can probably remove `ncurses-dev` from buildbot since we ended up not using it.

behlendorf commented:
Good call, will do.
